### PR TITLE
Add FTL base classes to runtimes-common.

### PR DESCRIPTION
**Pull request by dlorenc**
cc @mattmoor

I need to get linting and tests setup for this, just sending this out early so everyone is aware.

### CODE REVIEW

File: ftl/common/builder.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        info = tarfile.TarInfo(os.path.join('app', name))
+        info.size = len(content)
+        out.addfile(info, fileobj=cStringIO.StringIO(content))
+    
+    tar = buf.getvalue()
+    sha = 'sha256:' + hashlib.sha256(tar).hexdigest()
+
+    gz = cStringIO.StringIO()
+    with gzip.GzipFile(fileobj=gz, mode='w', compresslevel=1) as f:
+      f.write(tar)
```
Comment (mattmoor) [inline ftl/common/builder.py]: Can you factor a little library for [un]zipping a string, where we can put this?

This is something I've been wanting to do for `google/containerregistry` as a way of compartmentalizing the `gzip` problem and enabling us to experiment with fixes.

File: ftl/common/builder.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import os
+import tarfile
+
+from containerregistry.client.v2_2 import append
+
+
+import context
+
+
+class Base(object):
```
Comment (mattmoor) [inline ftl/common/builder.py]: Class comment

File: ftl/common/cache.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+import abc
+import hashlib
+
+from containerregistry.client import docker_name
+from containerregistry.client.v2_2 import docker_image
+from containerregistry.client.v2_2 import docker_session
+
+
+class Base(object):
```
Comment (mattmoor) [inline ftl/common/cache.py]: Class comment.

File: ftl/common/cache.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    """Lookup a cached image.
+    Args:
+      base_image: the docker_image.Image on which things are based.
+      namespace: a namespace for this cache.
+      checksum: the checksum of the package descriptor atop our base.
+      value: the docker_image.Image to store into the cache.
+    """
+
+
+class Registry(Base):
```
Comment (mattmoor) [inline ftl/common/cache.py]: class comment

File: ftl/common/cache.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    self._transport = transport
+    self._threads = threads
+    self._mount = mount or []
+
+  def __enter__(self):
+    return self
+
+  def _tag(self, base_image, namespace, checksum):
+    # TODO(mattmoor): pick up the latest containerregistry then use:
+    # fingerprint = '%s %s' % (base_image.digest(), checksum)
```
Comment (mattmoor) [inline ftl/common/cache.py]: You should be able to do this at HEAD of the repo.

File: ftl/common/context.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This package defines the interface for accessing the build context."""
+
+import abc
+import os
+
+
+class Base(object):
```
Comment (mattmoor) [inline ftl/common/context.py]: class comment

File: ftl/common/context.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+  # __enter__ and __exit__ allow use as a context manager.
+  @abc.abstractmethod
+  def __enter__(self):
+    """Initialize the context."""
+
+  def __exit__(self, unused_type, unused_value, unused_traceback):
```
Comment (mattmoor) [inline ftl/common/context.py]: weird that enter is abstract but this is not...

File: ftl/common/context.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  def GetFile(self, relative_path):
+    """Retrieve the contents of a particular file.
+    Args:
+      relative_path: The relative path of the file to read.
+    Returns:
+      the contents of the file.
+    """
+
+
+class Workspace(Base):
```
Comment (mattmoor) [inline ftl/common/context.py]: class comment

File: WORKSPACE [hunk 9]
```diff
@@ -32,12 +32,106 @@ new_go_repository(
 
 git_repository(
     name = "io_bazel_rules_docker",
+    commit = "a114402e4b98aee45ca67917bf7fb728913cf7d3",
     remote = "https://github.com/bazelbuild/rules_docker.git",
-    tag = "v0.0.1",
+)
+
+# For node-build
```
Comment (mattmoor) [inline WORKSPACE]: ftl?

File: ftl/common/cache.py (hunk 3 above)
Comment (dlorenc) [reply ftl/common/cache.py]: Done.

File: ftl/common/cache.py (hunk 4 above)
Comment (dlorenc) [reply ftl/common/cache.py]: done.

File: ftl/common/context.py (hunk 6 above)
Comment (dlorenc) [reply ftl/common/context.py]: Done.

File: WORKSPACE (hunk 9 above)
Comment (dlorenc) [reply WORKSPACE]: Done.

File: WORKSPACE [hunk 10]
```diff
@@ -32,12 +32,106 @@ new_go_repository(
...
+git_repository(
+    name = "subpar",
+    commit = "7e12cc130eb8f09c8cb02c3585a91a4043753c56",
+    remote = "https://github.com/google/subpar",
+)
+
+git_repository(
+    name = "containerregistry",
+    commit = "b0278a1544238d03648861b6d9395414d4c958e5",
+    remote = "https://github.com/google/containerregistry",
```
Comment (mattmoor) [inline WORKSPACE]: FWIW, I think most of the stuff above you should get for free from `rules_docker`.

File: WORKSPACE [hunk 11]
```diff
@@ -32,12 +32,106 @@ new_go_repository(
...
-  "docker_repositories"
+    "@io_bazel_rules_docker//docker:docker.bzl",
+    "docker_repositories",
+    "docker_pull",
 )
+
 docker_repositories()
+
+docker_pull(
+    name = "py_base",
```
Comment (mattmoor) [inline WORKSPACE]: py_base is nodejs, that deserves a comment :)

File: ftl/BUILD [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+load(
+    "@io_bazel_rules_docker//docker/contrib/python:image.bzl",
+    "py_image",
+)
+load(
+    "@io_bazel_rules_docker//docker:docker.bzl",
+    "docker_push",
+    "docker_build",
+)
+load("@subpar//:subpar.bzl", "par_binary")
```
Comment (mattmoor) [inline ftl/BUILD]: You don't need any of these `load` statements.

File: ftl/BUILD [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+py_library(
+    name = "ftl_lib",
+    srcs = glob([
+        "__init__.py",
+        "common/*.py",
```
Comment (mattmoor) [inline ftl/BUILD]: I think we want a `py_library` and corresponding `py_test` for each of the logic elements.  Not sure if you saw my other comment about adding unit tests?

Comment (dlorenc) [reply ftl/BUILD]: Yeah, working on tests now. I was planning on adding tests for everything that's not abstract, make sense?

As far as fakes, I'm not entirely sure what you had in mind. Something like a MemoryCache and a MemoryContext? I'm not sure what a fake builder would look like.

File: WORKSPACE (hunk 11 above)
Comment (dlorenc) [reply WORKSPACE]: D'oh, yeah this was

File: ftl/BUILD (hunk 13 above)
Comment (mattmoor) [reply ftl/BUILD]: Yeah, I think creating fakes on demand is probably fine.  IDK that a fake builder makes sense either.  Tests for everything instantiable SGTM.

**Comment by mattmoor** [comment]
@dlorenc Does what I'm doing here not work for you? https://github.com/bazelbuild/rules_k8s/pull/5

File: WORKSPACE (hunk 10 above)
Comment (dlorenc) [reply WORKSPACE]: Ah, nice.

File: ftl/BUILD (hunk 12 above)
Comment (dlorenc) [reply ftl/BUILD]: Cleaned.

File: ftl/common/cache_test.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        # The cache needs to calculate a digest from the base image.
+        fake_base = mock.MagicMock()
+        fake_base.digest.return_value = 'abc123'
+
+        mock_img = mock.MagicMock()
+        mock_from.return_value.__enter__.return_value = mock_img
+
+        # Test when the image exists.
+        mock_img.exists.return_value = True
+        c = cache.Registry('gcr.io/google-appengine', None, None)
```
Comment (mattmoor) [inline ftl/common/cache_test.py]: My paranoia wants you to change this to `fake.gcr.io` :)

File: ftl/common/cache_test.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+import mock
+
+
+class RegistryTest(unittest.TestCase):
+
+    def setUp(self):
+        pass
+
+    @mock.patch('containerregistry.client.v2_2.docker_image.FromRegistry')
```
Comment (mattmoor) [inline ftl/common/cache_test.py]: FWIW, a pattern I'm a big fan of is actually using `docker_build` to produce fake image data and then doing:
```python
with v2_image.FromTarball(...) as img:
   with mock.object.path(v2_image.FromRegistry, img):
      # Now FromRegistry will get "img" back.
```

File: ftl/common/cache_test.py (hunk 14 above)
Comment (dlorenc) [reply ftl/common/cache_test.py]: Done.

**Pull request merged**